### ISSUE TITLE
require tries to require from pkgs/rash/private

### ISSUE THREAD

**Issue by vityou** [willghatch/racket-rash#30]
If I have "test.rkt" in my home dir and I try to `require` it, I get:
```
#(struct:exn:fail:syntax:missing-module readline-input:16:9: cannot open module file
  module path: #<path:/home/vityou/.racket/6.12/pkgs/rash/private/test.rkt>
  path: /home/vityou/.racket/6.12/pkgs/rash/private/test.rkt
  system error: No such file or directory; errno=2 #<continuation-mark-set> (#<syntax:16:9 #<path:/home/vityou/.racket/6.1...>) /home/vityou/.racket/6.12/pkgs/rash/private/test.rkt)
```
I think it's because it evaluates `require` forms in the context of the rash/private/repl-namespace file.

**Comment by willghatch** [comment]
Yes, this is a known bug.  I talked with someone else about it a while ago, and there is a workaround:

```
(require racket/require-syntax)
(define-require-syntax (here stx)
  (syntax-parse stx
    [(_ path:str)
     (datum->syntax stx `(file ,(string-append (path->string (current-directory))
                                               (syntax->datum #'path))))]))
(require (here "file.rkt"))
```

or

```
(define-syntax (req-here stx)
  (syntax-parse stx
    [(_ path:str)
     (datum->syntax stx `(require
                          (file ,(string-append
                                  (path->string (current-directory))
                                  (syntax->datum #'path)))))]))
(req-here "file.rkt")
```

But this should have a solution.  I think the normal Racket REPL uses the current directory when using relative require paths, so I'm sure there's a way Rash can do it too.  I'm not sure what that is right now, though.

**Comment by willghatch** [comment]
I saw your mailing list question about this issue which provides the answer.  Thanks!  Do you want to send a pull request with the fix?  (I imagine you just need to add a line in `repl.rkt` adding the suggested parameterization.)

**Comment by willghatch** [comment]
I thought it would be as simple as adding the parameterization, but now I realize that it's probably not working because the namespace of the rash repl is from a module, so that is probably overriding the load path being parameterized.  I think we need to make the repl namespace be un-associated to a module like it currently is.  I'll make that change in the near-ish future, unless you want to take a crack at it sooner.

**Comment by vityou** [comment]
I wouldn't say it's super important.  I mainly tried to find a solution because I thought it would be easy to fix.  The `here` workaround is fine until you fix it in the near future, or until I learn more about racket and rash to fix it.  Currently I'm reading fear of macros, and I don't have much knowledge on namespaces and anchors, so it may be a while before I can contribute anything non superficial to rash.

**Comment by willghatch** [comment]
Only a few issues like this in the repl would require any knowledge of namespaces.  Macro knowledge is needed to make interesting pipeline operators and line macros.  But a lot of stuff needed to improve Rash will be writing normal functions.  I want to have useful commands written as Racket functions (I've been thinking about good conventions for types of objects for them to pass, working with the generic system, etc), once there is an editor I want editing functions, ...

I think outside the core infrastructure of Rash there is a lot to build around it to make it into a great system that won't require as much specialized knowledge about esoteric topics.  But I definitely recommend Fear of Macros and other macro tutorials, because macros are amazing and the core technology behind Rash and most cool things in Racket.

**Comment by willghatch** [comment]
That is to say, I'm happy for the things you've contributed so far, and I think you'll be able to contribute useful things more quickly than you think.  Especially, one very useful thing is simply another point of view to see things that I don't and to give feedback.

**Comment by vityou** [comment]
After reading a bit about namespaces, it seems like most of the stuff in the `repl-namespace.rkt` file could be replaced with

```
(provide repl-namespace)
(define repl-namespace (make-base-namespace))

(parameterize ([current-namespace repl-namespace])
  (namespace-require whatever) ; though I'm not sure how to require relative paths
  (eval '(define-macro/function ...)
  (eval ...)))
```

If that's all there is to it, I could probably take a shot at it.

**Comment by vityou** [comment]
I have made some progress on it, pretty much everything works.  [Here](https://github.com/vityou/racket-rash/blob/fix-namespace/rash/repl.rkt) is what I have so far.  The only issues I ran into was that when you had the `eval-syntax` forms that evaluated something in the namespace, they didn't work with the new namespace not being associated with a file, so I just turned them into `eval`s.  

The only thing is that now there is a 
```
Result n:
#<pipeline>
```
displayed after everything, which I can't seem to fix.  It may have to do with [this](https://github.com/vityou/racket-rash/blob/285503db247a7beb1d06dcbe40c41697709fedf7/rash/repl.rkt#L76), since it was main thing I changed, but I'm not sure.

**Comment by willghatch** [comment]
I was also looking at this a bit earlier and came to a similar situation.  There is a shallow problem here and a deeper problem as well.  The shallow problem is that `#<pipeline>` is printed because the default line macro in the repl makes it automatically adds `&pipeline-ret`.  This is actually stupid and I've been meaning to make it just the normal `run-pipeline` or `run-pipeline/logic` rather than the custom one I set.  

The bigger problem is that pipeline objects are no longer being recognized by the `top-level-print-format` function, which I added to be able to massage how things print -- especially pipelines, since I was doing this stupid thing and always returning them.

So basically it looks like when you require modules dynamically in a namespace, the struct definitions are seemingly not the same as from the same modules that were required (perhaps transitively) by the program doing `eval` in a namespace.  This seems strange, though, because top-level forms (eg. print expressions at the top-level) of a module that is required by both the program and the namespace are only evaluated once.

Anyway, I'm going to figure out what is going on with that before I move forward with anything here.

**Comment by vityou** [comment]
Hmm, I'm just guessing, but maybe it would work if `top-level-print-format` evaluated it's tests in the context of the `repl-namespace`.  Because when I'm in the repl, and type `(pipeline? (hash-ref interactive-return-values 1))`, (where the last return value was an empty pipeline) it returns true.  Sort of a hacky solution (if it works), since it doesn't fix `top-level-print-format` not recognizing them.

**Comment by vityou** [comment]
I can confirm it's not just rash, if you have these files in the same directory:
```
; struct-exporter.rkt
#lang racket

(provide (struct-out foo))

(struct foo (f1 f2))
```
and
```
; foo.rkt
#lang racket

(require "struct-exporter.rkt")

(define ns (make-base-namespace))

(parameterize ([current-namespace ns]
               [current-directory "/path/to/dir"])
  (namespace-require "struct-exporter.rkt"))

(foo? (parameterize ([current-namespace ns])
  (eval '(foo 1 2))))
```
The `(foo? ...)` evaluates to false.

**Comment by willghatch** [comment]
Well, I've pushed some changes that fix things.  It now loads the formatter function through the namespace, so the predicates work.  But also I took out the pipeline-return stuff and then simplified the formatter function.  So now things are basically like they were before, except now the load path works correctly.

Also I added some simple command-line parsing to allow readline to be enabled and disabled.
